CIP-2551/CIP-2551.md

```diff
@@ -0,0 +1,271 @@
+have efficient hashes both in-circuit and on-chain.
+
+More generally, we can decouple the choice of zk-SNARK theory for on-chain
+proof verification from the choice of zk-SNARK theory for off-chain
+recursive proof building, as long as we can reasonably efficiently
+convert one form of zk-SNARK to the other using recursive proving.
+
+For efficient off-chain proving, it makes sense to use a FRI-based
+zk-SNARK theory, since this leads to relatively fast proving times,
+at the cost of larger proof sizes, compared to Halo 2 for example. This
+also has the advantage that recursive proofs are relatively simpler to
+implement, because they do not require an accumulation scheme over a
+2-cycle of curves. FRI allows for homogeneous recursive proving.
+
+For efficient on-chain proof verification, it would make sense to use
+an elliptic curve based zk-SNARK theory, such as Halo 2, since this
+leads to relatively small proof sizes, potentially allowing for a proof
+to verified in a single Cardano transaction.  However, realizing this
+possibility requires primitives for a suitable choice of elliptic curve,
+one which is amenable to efficient proof verification.
+
+For realizing Halo 2 proof verification on-chain, we would want primitives
+for Pasta curves. However, Pasta curves are not the most ideal choice of
+curve for proofs that will not be recursively input to other proofs. When
+this is not required, the Ed25519 curve is a better choice because using
+it reduces proof size and verification costs. When using FRI for off-chain
+recursive proof composition, this requirement is not present. Instead, we
+would have a zk-SNARK over an elliptic curve which verifies a FRI-based
+zk-SNARK, and that zk-SNARK over the elliptic curve would be the final
+proof in the tree of recursive proofs, the one which is posted and
+verified on-chain.
+
+On-chain Halo 2 proof verification would open the door to on-chain
+verification of more or less all zk-SNARKs, and not just those
+mentioned. Halo 2 is a flexible enough theory to encode computable
+functions and thus the verification algorithms of arbitrary zk-SNARK
+theories. Therefore this CIP should be seen as a step towards efficient
+on-chain verification of zk-SNARKs in general.
+
+That describes our particular motivation for wanting the Ed25519 curve in
+Plutus at Orbis Labs. Furthermore, we believe that the Ed25519 curve is a
+broadly useful cryptographic primitive, when one needs a cryptographically
+secure elliptic curve but one does not need a pairing-friendly curve or
+a 2-cycle of curves.
+
+
+## Specification
+
+Add to Plutus the following types:
+
+ * `Ed25519GElement`: the type of elements of the Ed25519 curve.
+ * `Ed25519FElement`: the type of elements of the finite field of order
+    2^255 - 19.
+
+Add the following functions:
+
+ * Basic group operations:
+    * `Ed25519_G_add :: Ed25519GElement -> Ed25519GElement -> Ed25519GElement`
+    * `Ed25519_G_neg :: Ed25519GElement -> Ed25519GElement`
```

Comment:
These functions could fail, right? So, shouldn't they use a Maybe? Above, you mention that the `Ed25519GElement` is an Ed25519 curve element, but in practices this is just a wrapper of two field elements, it could be that one or both of these tuples do not lie on the curve.